pkg/runtime.v2/core/trainingruntime_test.go

```diff
@@ -53,59 +67,212 @@ func TestTrainingRuntimeNewObjects(t *testing.T) {
+				testingutil.MakeTrainingRuntimeSpecWrapper(testingutil.MakeTrainingRuntimeWrapper(metav1.NamespaceDefault, "test-runtime").Spec).
+					TorchPolicy(100, "auto").
+					ContainerTrainer("test:runtime", []string{"runtime"}, []string{"runtime"}, resRequests).
+					Obj(),
+			).Obj(),
 			trainJob: testingutil.MakeTrainJobWrapper(metav1.NamespaceDefault, "test-job").
 				UID("uid").
 				RuntimeRef(kubeflowv2.SchemeGroupVersion.WithKind(kubeflowv2.TrainingRuntimeKind), "test-runtime").
 				Trainer(
 					testingutil.MakeTrainJobTrainerWrapper().
-						ContainerImage("test:trainjob").
+						NumNodes(30).
+						NumProcPerNode("3").
+						Obj(),
+				).
+				Obj(),
+			wantObjs: []client.Object{
+				testingutil.MakeJobSetWrapper(metav1.NamespaceDefault, "test-job").
+					NumNodes(30).
+					ContainerTrainer("test:runtime", []string{"runtime"}, []string{"runtime"}, resRequests).
+					ContainerTrainerEnv(
+						[]corev1.EnvVar{
+							{
+								Name:  constants.TorchEnvNumNodes,
+								Value: "30",
+							},
+							{
+								Name:  constants.TorchEnvNumProcPerNode,
+								Value: "3",
+							},
+						},
+					).
+					ControllerReference(kubeflowv2.SchemeGroupVersion.WithKind(kubeflowv2.TrainJobKind), "test-job", "uid").
+					Obj(),
+			},
+		},
+		"succeeded to build JobSet with Torch values from the Runtime and envs.": {
+			trainingRuntime: testingutil.MakeTrainingRuntimeWrapper(metav1.NamespaceDefault, "test-runtime").RuntimeSpec(
+				testingutil.MakeTrainingRuntimeSpecWrapper(testingutil.MakeTrainingRuntimeWrapper(metav1.NamespaceDefault, "test-runtime").Spec).
+					TorchPolicy(100, "auto").
+					ContainerTrainer("test:runtime", []string{"runtime"}, []string{"runtime"}, resRequests).
+					ContainerTrainerEnv(
+						[]corev1.EnvVar{
+							{
+								Name:  "TRAIN_JOB",
+								Value: "original",
+							},
+							{
+								Name:  "RUNTIME",
+								Value: "test:runtime",
+							},
+						},
+					).
+					Obj(),
+			).Obj(),
+			trainJob: testingutil.MakeTrainJobWrapper(metav1.NamespaceDefault, "test-job").
+				UID("uid").
+				RuntimeRef(kubeflowv2.SchemeGroupVersion.WithKind(kubeflowv2.TrainingRuntimeKind), "test-runtime").
+				Trainer(
+					testingutil.MakeTrainJobTrainerWrapper().
+						ContainerTrainer("test:trainjob", []string{"trainjob"}, []string{"trainjob"}, resRequests).
+						ContainerTrainerEnv(
+							[]corev1.EnvVar{
+								{
+									Name:  "TRAIN_JOB",
+									Value: "override",
+								},
+								{
+									Name:  "TRAIN_JOB_CUSTOM",
+									Value: "test:trainjob",
+								},
+							},
+						).
+						Obj(),
+				).
+				Obj(),
+			wantObjs: []client.Object{
+				testingutil.MakeJobSetWrapper(metav1.NamespaceDefault, "test-job").
+					NumNodes(100).
+					ContainerTrainer("test:trainjob", []string{"trainjob"}, []string{"trainjob"}, resRequests).
+					ContainerTrainerEnv(
+						[]corev1.EnvVar{
```

Comment:
@tenzen-y Do you know why these cmpOpt doesn't work?
https://github.com/kubeflow/training-operator/blob/60a452b2dd33ee5f16f2c1cd79b74d8cd2219ad7/pkg/runtime.v2/core/trainingruntime_test.go#L282-L288

I noticed that this unit test is flaky given that we iterate over map to populate container env variables, and order can't be guaranteed :
https://github.com/kubeflow/training-operator/blob/60a452b2dd33ee5f16f2c1cd79b74d8cd2219ad7/pkg/runtime.v2/framework/plugins/jobset/builder.go#L83-L91

Reply:
Actually, thought on this a little bit more. We can't use Map here given that order can't be guaranteed while iterating over map.
Since we [compare old and new JobSet here](https://github.com/kubeflow/training-operator/blob/master/pkg/runtime.v2/framework/plugins/jobset/jobset.go#L122), we need to make sure that we keep list of envs consistent during Build() stage.

Reply:
We need to specify the dedicated typed sorting functions like this:

```go
cmpopts.SortMaps(func(a, b corev1.EnvVar) bool { return a.Name < b.Name }),
```

Reply:
Yes, this could be one of the options to improve and refactor the code.
@tenzen-y Do you want me to create TODO for it in the Info object to use sorted Map ?
The benefit would be to not have additional var as I use here: https://github.com/kubeflow/training-operator/blob/85d0cf1fca0e56d206e579f77d68f33aca1a0f87/pkg/runtime.v2/framework/plugins/torch/torch.go#L100-L103

Reply:
Yeah, let's leave a TODO comment.